### ISSUE TITLE
Universe polymorphism breaks well-foundedness proofs of program definitions

### ISSUE THREAD

**Issue by Tuplanolla** [coq/coq#13979]
#### Version

Coq 8.12.2

#### Operating system

Red Hat Enterprise Linux Workstation 7.4

#### Description of the problem

The following proof does not go through
in the presence of universe polymorphism,
which is questionable.

```coq
From Coq Require Import Lia Lists.List Program.Wf.

Import ListNotations Nat.

Set Universe Polymorphism.

Program Fixpoint merge_by {A : Type} (f : A -> nat) (l0 l1 : list A)
  {measure (length l0 + length l1)} : list A :=
  match l0, l1 with
  | [], _ => l1
  | _, [] => l0
  | n0 :: k0, n1 :: k1 => if f n0 <=? f n1 then
    n0 :: merge_by f k0 l1 else
    n1 :: merge_by f l0 k1
  end.
Next Obligation.
  intros.
  subst;cbn.
  lia.
Qed.
```

```
Illegal application:
The term "merge_by_func" of type
 "forall recarg : {A : Type & {_ : A -> nat & {_ : list A & list A}}},
  let A := projT1 recarg in
  let f := projT1 (projT2 recarg) in
  let l0 := projT1 (projT2 (projT2 recarg)) in let l1 := projT2 (projT2 (projT2 recarg)) in list A"
cannot be applied to the term
 "existT (fun A : Type => {_ : A -> nat & {_ : list A & list A}}) A
    (existT (fun _ : A -> nat => {_ : list A & list A}) f (existT (fun _ : list A => list A) l0 l1))"
   : "{A : Type & {_ : A -> nat & {_ : list A & list A}}}"
This term has type "{A : Type@{Top.1} & {_ : A -> nat & {_ : list A & list A}}}" which should be coercible to
 "{A : Type@{Top.2} & {_ : A -> nat & {_ : list A & list A}}}".
```

#### Notes

This is in conflict with #13980 and may be related to #8432 and #13765.

**Comment by SkySkimmer** [comment]
On 8.11.2 I get uncaught Not_found on Qed
On master I get 
~~~
Error:
<in exception printer>:
Raised at file "clib/int.ml", line 39, characters 14-29
Called from file "clib/hMap.ml", line 365, characters 12-28
Called from file "library/nametab.ml", line 554, characters 17-64
Called from file "interp/constrextern.ml" (inlined), line 313, characters 35-62
Called from file "interp/constrextern.ml", line 963, characters 4-31
Called from file "interp/constrextern.ml", line 987, characters 71-95
Called from file "printing/proof_diffs.ml", line 253, characters 37-96
Called from file "printing/printer.ml", line 98, characters 3-37
Called from file "vernac/himsg.ml", line 98, characters 36-55
Called from file "vernac/himsg.ml", line 389, characters 15-44
Called from file "vernac/himsg.ml", line 1410, characters 11-16
Called from file "lib/cErrors.ml", line 100, characters 10-13
Called from file "lib/cErrors.ml", line 113, characters 4-49
<original exception:
Anomaly "Uncaught exception Not_found."
Please report at http://coq.inria.fr/bugs/.
Raised at file "kernel/type_errors.ml", line 137, characters 2-59
Called from file "kernel/typeops.ml", line 507, characters 14-49
Called from file "kernel/typeops.ml", line 513, characters 21-36
Called from file "kernel/typeops.ml", line 513, characters 21-36
Called from file "kernel/typeops.ml", line 513, characters 21-36
Called from file "kernel/typeops.ml", line 513, characters 21-36
Called from file "kernel/typeops.ml", line 644, characters 18-36
Called from file "kernel/term_typing.ml", line 169, characters 14-36
Called from file "kernel/term_typing.ml", line 329, characters 4-30
Called from file "kernel/safe_typing.ml", line 905, characters 8-53
Called from file "library/global.ml", line 72, characters 16-31
Called from file "vernac/declare.ml", line 358, characters 11-54
Called from file "vernac/declare.ml", line 364, characters 11-49
Called from file "vernac/comProgramFixpoint.ml", line 242, characters 16-113
Called from file "vernac/declare.ml", line 582, characters 2-52
Called from file "vernac/declare.ml", line 686, characters 2-86
Called from file "vernac/declare.ml", line 1056, characters 17-79
Called from file "vernac/declare.ml", line 1158, characters 19-46
Called from file "vernac/declare.ml", line 1187, characters 28-78
Called from file "vernac/declare.ml", line 1243, characters 4-79
Called from file "vernac/vernacinterp.ml", line 239, characters 18-75
Called from file "lib/flags.ml", line 17, characters 14-17
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "vernac/vernacinterp.ml", line 264, characters 18-43
Called from file "vernac/vernacinterp.ml", line 262, characters 6-279
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "stm/stm.ml", line 2266, characters 20-71
Called from file "stm/stm.ml", line 957, characters 6-10
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "stm/stm.ml", line 2296, characters 4-105
Called from file "stm/stm.ml", line 2397, characters 4-60
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "toplevel/vernac.ml", line 68, characters 31-52
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "toplevel/coqloop.ml", line 400, characters 17-62
Called from file "toplevel/coqloop.ml", line 437, characters 4-39
Called from file "toplevel/coqloop.ml", line 443, characters 6-34
~~~

**Comment by Tuplanolla** [comment]
Looks like `Not_found` also comes up from `Program Fixpoint` in #13945.

**Comment by SkySkimmer** [comment]
That's clearly a different issue.

**Comment by SkySkimmer** [comment]
It seems coqtop gets Not_found and coqc gets a regular error.

Simplified script:
~~~coq
From Coq Require Import Program.Wf.

Open Scope list_scope.

Set Universe Polymorphism.

Obligation Tactic := idtac.

Axiom admit : forall P : Prop, P.

Program Fixpoint merge_by {A : Type} (f : A -> nat) (l0 l1 : list A)
  {measure (length l0 + length l1)} : list A :=
  match l0, l1 with
  | nil, _ => l1
  | _, nil => l0
  | n0 :: k0, n1 :: k1 => merge_by f k0 l1
  end.

Next Obligation.
  apply admit.
Qed.

Next Obligation.
  apply admit.
Qed.

Next Obligation.
  apply admit.
Qed.
~~~
same error on Qed

**Comment by herbelin** [comment]
Fixed in #18918
